llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp

```diff
@@ -943,16 +943,29 @@ bool CompileTimePropertiesPass::transformSYCLPropertiesAnnotation(
     LLVMContext &Ctx = M.getContext();
     unsigned MDKindID = Ctx.getMDKindID(SpirvDecorMdKind);
     if (!FPGAProp && llvm::isa<llvm::Instruction>(IntrInst->getArgOperand(0))) {
-      // If there are no annotations other than cache controls we can apply the
-      // controls to the pointer and remove the intrinsic.
+      // Find all load/store instructions using the pointer being annotated and
+      // apply the cache control metadata to them.
+      SmallVector<std::pair<Instruction *, int>, 8> TargetedInstList;
+      getUserListIgnoringCast<LoadInst>(IntrInst, TargetedInstList);
+      getUserListIgnoringCast<StoreInst>(IntrInst, TargetedInstList);
+      getUserListIgnoringCast<MemTransferInst>(IntrInst, TargetedInstList);
+      for (const auto &Pair : TargetedInstList) {
+        auto *Inst = Pair.first;
+        // Merge with existing metadata if present.
+        SmallVector<Metadata *, 8> MDOps;
+        if (MDNode *CurrentMD = Inst->getMetadata(MDKindID))
+          for (Metadata *Op : CurrentMD->operands())
+            MDOps.push_back(Op);
+        for (Metadata *Op : MDOpsCacheProp)
+          MDOps.push_back(Op);
+        MDOps.push_back(ConstantAsMetadata::get(Constant::getIntegerValue(
+            Type::getInt32Ty(Ctx), APInt(32, Pair.second))));
+        Inst->setMetadata(MDKindID, MDTuple::get(Ctx, MDOps));
+      }
```

Comment:
Small stylistic preference:
```suggestion
      for (const auto &[Inst, MDVal] : TargetedInstList) {
        // Merge with existing metadata if present.
        SmallVector<Metadata *, 8> MDOps;
        if (MDNode *CurrentMD = Inst->getMetadata(MDKindID))
          for (Metadata *Op : CurrentMD->operands())
            MDOps.push_back(Op);
        for (Metadata *Op : MDOpsCacheProp)
          MDOps.push_back(Op);
        MDOps.push_back(ConstantAsMetadata::get(Constant::getIntegerValue(
            Type::getInt32Ty(Ctx), APInt(32, MDVal))));
        Inst->setMetadata(MDKindID, MDTuple::get(Ctx, MDOps));
      }
```